### PR TITLE
Code refactoring and code cleanup of Satellite.py and task.py script.

### PR DESCRIPTION
**Pull request by devendra104**
#366

### CODE REVIEW

File: upgrade/helpers/tasks.py [hunk 1]
```diff
@@ -540,30 +540,12 @@ def setup_foreman_maintain():
...
-        ))
-        tools_repo.write('enabled=1\n')
-        tools_repo.write('gpgcheck=0\n')
-        put(local_path=tools_repo,
-            remote_path='/etc/yum.repos.d/sat6tools7.repo')
-        tools_repo.close()
+        repository_setup("[sat6]", "satellite 6", base_url, 1, 0)
+        tools_url = "{}".format(os.environ.get('TOOLS_RHEL7'))
+        repository_setup("[sat6tools7]", "satellite6-tools7", tools_url, 1, 0)
+
```
Comment (vijay8451) [inline upgrade/helpers/tasks.py]: can we use above new changes in below way:
```
        repository_setup("[sat6]",
                         "satellite 6",
                         '{}'.format(os.environ.get('BASE_URL')),
                         1, 0)
        repository_setup("[sat6tools7]",
                         "satellite6-tools7",
                         '{}'.format(os.environ.get('TOOLS_RHEL7')),
                         1, 0)
```

File: upgrade/helpers/tasks.py [hunk 2]
```diff
@@ -587,17 +569,9 @@ def setup_foreman_maintain_repo():
...
-            os.environ.get('MAINTAIN_REPO')
-        ))
-        maintain_repo.write('enabled=1\n')
-        maintain_repo.write('gpgcheck=0\n')
-        put(local_path=maintain_repo,
-            remote_path='/etc/yum.repos.d/foreman-maintain.repo')
-        maintain_repo.close()
+        maintain_repo_url = "{}".format(os.environ.get('MAINTAIN_REPO'))
+        repository_setup("[foreman-maintain]", "foreman-maintain",
+                         maintain_repo_url, 1, 0)
```
Comment (vijay8451) [inline upgrade/helpers/tasks.py]: same as above.

File: upgrade/helpers/tasks.py [hunk 3]
```diff
@@ -672,17 +646,9 @@ def upgrade_puppet3_to_puppet4():
...
-            os.environ.get('PUPPET4_REPO')
-        ))
-        satellite_repo.write('enabled=1\n')
-        satellite_repo.write('gpgcheck=0\n')
-        put(local_path=satellite_repo,
-            remote_path='/etc/yum.repos.d/puppet4.repo')
-        satellite_repo.close()
+        puppet_repo_url = "{}".format(os.environ.get('PUPPET4_REPO'))
+        repository_setup("[Puppet4]", "puppet4",
+                         puppet_repo_url, 1, 0)
```
Comment (vijay8451) [inline upgrade/helpers/tasks.py]: same as above .

File: upgrade/helpers/tasks.py [hunk 4]
```diff
@@ -737,17 +696,9 @@ def setup_satellite_clone():
...
-            os.environ.get('MAINTAIN_REPO')
-        ))
-        satellite_repo.write('enabled=1\n')
-        satellite_repo.write('gpgcheck=0\n')
-        put(local_path=satellite_repo,
-            remote_path='/etc/yum.repos.d/maintain.repo')
-        satellite_repo.close()
+        maintain_repo_url = "{}".format(os.environ.get('MAINTAIN_REPO'))
+        repository_setup("[maintainrepo]", "maintain",
+                         maintain_repo_url, 1, 0)
```
Comment (vijay8451) [inline upgrade/helpers/tasks.py]: same here .

File: upgrade/helpers/tasks.py (hunk 3 above)
Comment (vijay8451) [inline upgrade/helpers/tasks.py]: same as above .

File: upgrade/satellite.py [hunk 5]
```diff
@@ -74,189 +71,150 @@ def satellite6_setup(os_version):
...
+            TO_VERSION
+                Satellite version to upgrade to and enable repos while upgrading.
+                e.g '6.1','6.2', '6.3'
+            PERFORM_FOREMAN_MAINTAIN_UPGRADE
+                use foreman-maintain for satellite upgrade
+
+        - Upgrades Satellite Server to its latest zStream version
+            Note: For zstream upgrade both 'To' and 'From' version should be same
+
+            FROM_VERSION
```
Comment (jyejare) [inline upgrade/satellite.py]: We are missing FROM_VERSION from old version docstring.

File: upgrade/satellite.py [hunk 6]
```diff
@@ -74,189 +71,150 @@ def satellite6_setup(os_version):
     return sat_host
 
 
-def satellite6_upgrade():
-    """Upgrades satellite from old version to latest version.
+def satellite6_upgrade(upgrade_type=None):
```
Comment (jyejare) [inline upgrade/satellite.py]: Docstring is missing for upgrade_type parameter.

File: upgrade/satellite.py [hunk 7]
```diff
@@ -74,189 +71,150 @@ def satellite6_setup(os_version):
...
-        Satellite version to upgrade to and enable repos while upgrading.
-        e.g '6.1','6.2', '6.3'
-    PERFORM_FOREMAN_MAINTAIN_UPGRADE
-        use foreman-maintain for satellite upgrade
     """
     logger.highlight('\n========== SATELLITE UPGRADE =================\n')
     to_version = os.environ.get('TO_VERSION')
+    from_version = os.environ.get('FROM_VERSION')
+    if upgrade_type:
+        if not from_version == to_version:
```
Comment (jyejare) [inline upgrade/satellite.py]: Upgrade will always fail for old to new upgrade as the versions will be different always.

File: upgrade/satellite.py [hunk 8]
```diff
@@ -74,189 +71,150 @@ def satellite6_setup(os_version):
...
 
-    Note: For zstream upgrade both 'To' and 'From' version should be same
 
-    FROM_VERSION
-        Current satellite version which will be upgraded to latest version
-    TO_VERSION
-        Next satellite version to which satellite will be upgraded
-    PERFORM_FOREMAN_MAINTAIN_UPGRADE
-        use foreman-maintain for satellite upgrade
+def enable_disable_repo(enable_repo_name, disable_repo_name):
```
Comment (jyejare) [inline upgrade/satellite.py]: Move this to tasks ?

File: upgrade/satellite.py [hunk 9]
```diff
@@ -74,189 +71,150 @@ def satellite6_setup(os_version):
...
+    :param disable_repo_name: This will take the
+    list of repository which you are going to disable
+    """
+    if disable_repo_name:
+        [disable_repos('{}'.format(repo), silent=True) for repo in disable_repo_name]
+    if enable_repo_name:
+        [enable_repos('{}'.format(repo)) for repo in disable_repo_name]
+
+
+def repository_cleanup(repo_name):
```
Comment (jyejare) [inline upgrade/satellite.py]: Move this to tasks?

File: upgrade/satellite.py [hunk 10]
```diff
@@ -74,189 +71,150 @@ def satellite6_setup(os_version):
...
+    The purpose of this function to perform the repository cleanup on the basis of
+    their name.
+    :param repo_name: repository name
+    """
+    for fname in os.listdir('/etc/yum.repos.d/'):
+        if repo_name in fname.lower():
+            os.remove('/etc/yum.repos.d/{}'.format(fname))
+
+
+def upgrade_task():
```
Comment (jyejare) [inline upgrade/satellite.py]: better naming is required something like manual_upgrade or nonfm_upgrade

File: upgrade/satellite.py (hunk 7 above)
Comment (devendra104) [reply upgrade/satellite.py]: Correct, this part will work only when upgrade_type is zStream.

File: upgrade/satellite.py (hunk 10 above)
Comment (devendra104) [reply upgrade/satellite.py]: Agree, I am keeping nonfm_upgrade.

File: upgrade/satellite.py (hunk 9 above)
Comment (devendra104) [reply upgrade/satellite.py]: Moved.

File: upgrade/satellite.py (hunk 8 above)
Comment (devendra104) [reply upgrade/satellite.py]: Moved

File: upgrade/satellite.py (hunk 5 above)
Comment (devendra104) [reply upgrade/satellite.py]: Added FROM_VERSION.

File: upgrade/satellite.py (hunk 6 above)
Comment (devendra104) [reply upgrade/satellite.py]: Incorporated.

File: upgrade/helpers/tasks.py [hunk 11]
```diff
@@ -887,3 +845,25 @@ def check_status_of_running_task(command, attempt):
...
+    """
+    satellite_repo = StringIO()
+    satellite_repo.write('{}\n'.format(repository))
+    satellite_repo.write('name=s{}\n'.format(repository_name))
+    satellite_repo.write('baseurl={0}\n'.format(base_url))
+    satellite_repo.write('enabled={}\n'.format(enable))
+    satellite_repo.write('gpgcheck={}\n'.format(gpgcheck))
+    put(local_path=satellite_repo,
+        remote_path='/etc/yum.repos.d/{}.repo'.format(repository.
+                                                      replace("[", '').replace("]", '')))
```
[the remark this answers is not in the document]
Comment (devendra104) [reply upgrade/helpers/tasks.py]: My bad, I didn't think about it.

**Comment by devendra104** [comment]
The first phase of testing has been done, Satellite upgraded successfully from one version to other with the latest changes.
Pending item:
zStream testing is pending.
Testing with the latest snap.

**Comment by devendra104** [comment]
6.6 Satellite upgrade successfully with the latest code.

**Comment by devendra104** [comment]
Pending item:
zStream testing

**Comment by devendra104** [comment]
I have re-tested again with the foreman-maintain upgrade and normal upgrade, both the task completed successfully.

**Comment by devendra104** [comment]
zStream testing is pending and it won't be possible in this release.

File: upgrade/helpers/tasks.py [hunk 12]
```diff
@@ -887,3 +855,71 @@ def check_status_of_running_task(command, attempt):
         logger.info("Command: {}\n"
                     "Exceeded the maximum attempt to check the "
                     "tasks running status".format(command))
+
+
+def repository_setup(repository, repository_name, base_url, enable, gpgcheck):
+    """
+    This is generic fucntion which is used to setup the repository
+    :param repository: uniq repository ID
```
Comment (jyejare) [inline upgrade/helpers/tasks.py]: Rename all docstrings as below:
```suggestion
    :param str repository: uniq repository ID
```
The parameter type in docstring is missing.

File: upgrade/satellite.py [hunk 13]
```diff
@@ -74,189 +72,110 @@ def satellite6_setup(os_version):
...
-    TO_VERSION
-        Satellite version to upgrade to and enable repos while upgrading.
-        e.g '6.1','6.2', '6.3'
-    PERFORM_FOREMAN_MAINTAIN_UPGRADE
-        use foreman-maintain for satellite upgrade
     """
     logger.highlight('\n========== SATELLITE UPGRADE =================\n')
     to_version = os.environ.get('TO_VERSION')
+    from_version = os.environ.get('FROM_VERSION')
+    if upgrade_type:
```
Comment (jyejare) [inline upgrade/satellite.py]: Rename this to zStream and set this by default False.

Comment (jyejare) [reply upgrade/satellite.py]: Rename this parameter to zStream everywhere.

File: upgrade/helpers/tasks.py (hunk 12 above)
Comment (devendra104) [reply upgrade/helpers/tasks.py]: good observation :+1:

**Pull request merged**